### PR TITLE
Upgrade jsonfield from 2.1.1 to 3.1.0

### PR DESCRIPTION
**Pull request by gherceg**
## Technical Summary

https://dimagi-dev.atlassian.net/browse/SAAS-15191

v2.1.1 of jsonfield uses [ugettext_lazy](https://github.com/rpkilby/jsonfield/blob/2.1.1/jsonfield/fields.py#L8) which was removed [v3.0.0](https://github.com/rpkilby/jsonfield/blob/3.0.0/src/jsonfield/fields.py#L6). Django no longer supports this method, and instead expects gettext_lazy to be used in its place (this is a blocking issue for upgrading to django 4.2).

As far as the jsonfield upgrade itself, it's a bit of a strange [changelog](https://github.com/rpkilby/jsonfield/blob/master/CHANGES.rst) since the forked jsonfield2 was brought back into jsonfield as part of the v3 release, but the v2.1.1 changelog was removed. ~After looking over that, it doesn't seem like anything will impact us, but I'm also not 100% sure that is true, especially given how widely used this field is in our code. I'll deploy this to staging to see if anything obvious pops up.~

### TypeError: the JSON object must be str, bytes or bytearray, not dict
The first issue encountered was the failing test `corehq.messaging.scheduling.tests.test_views:MessagingImageUploadTest.test_delete_unused_messaging_images
` raising the error `TypeError: the JSON object must be str, bytes or bytearray, not dict`. The problematic code was
```
    email_messages = [
        json.loads(message)
        for message in (EmailContent.objects
                        .values_list("html_message", flat=True)
                        .filter(html_message__isnull=False))
    ]
```
In v2.1.1, `message` was a dictionary serialized into a string, so `json.loads(message)` did not raise any error.
```
'{"*": "<img src=\\"http://localhost:8000/a/fake/messaging/image/download/7v7BHMcwBLp1SRlEQoionA\\" />"}'
```
In v3.1.0, `message` is returned as a pure dictionary, so `json.loads(message)` is an invalid call. When not annotating the query with `values_list`, `message` is a dictionary in both versions, but `values_list` was causing the difference. Looking closer, this [from_db_value](https://github.com/rpkilby/jsonfield/blob/3.1.0/src/jsonfield/fields.py#L50-L58) method was added to JSONField, which django is aware of and [uses](https://github.com/django/django/blob/3.2.23/django/db/models/sql/compiler.py#L1132-L1138) when returning results. That method relies on [checked_loads](https://github.com/rpkilby/jsonfield/blob/3.1.0/src/jsonfield/json.py) which will deserialize the dictionary, ultimately returning a dictionary instead of a string as it did before.

TL;DR: we no longer need to explicitly call `json.loads(message)` in our code, because `JSONField.from_db_value` handles this for us.

### Rollback instructions

Once on django 4.2, reverting this would cause a runtime error.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by gherceg** [comment]
@millerdev I spent awhile debugging the failing test that I updated in e63a7d8438965873a7cc64aef8ebab99766bee62, thinking I would be able to point to a specific change in jsonfield that led to the `properties` field returning None instead of an empty dictionary after fetching the saved `BlobMeta` from the db. However I ended up surprised that it ever returned an empty dictionary once we fetched the saved version of `BlobMeta` due to the logic in [NullJsonField](https://github.com/dimagi/commcare-hq/blob/master/corehq/util/models.py#L237-L240) which ensures a value of None is written to the database if the dictionary is empty. Any ideas why the empty dictionary was still being returned previously?

**Review by millerdev — APPROVED** [review]
I took the liberty of pushing a few new commits, which I think fixes the issue in a more correct way. I approve of all but the last commit you made (which I reverted). Feel free to merge after reviewing if you like my changes.

File: corehq/blobs/tests/test_metadata.py [hunk 1]
```diff
@@ -54,11 +54,13 @@ def test_save_properties(self):
...
         self.assertEqual(get_meta(meta).properties, {"mood": "Vangelis"})
 
-    def test_save_empty_properties(self):
+    def test_save_empty_properties_returns_none(self):
         meta = new_meta()
         self.assertEqual(meta.properties, {})
+
         self.db.put(BytesIO(b"content"), meta=meta)
-        self.assertEqual(get_meta(meta).properties, {})
+        self.assertIsNone(get_meta(meta).properties)
```
[the remark this answers is not in the document]
Comment (gherceg) [reply corehq/blobs/tests/test_metadata.py]: Ah I see. I definitely looked closely at `from_db_value` so a bit bummed I didn't put this all together. I'm surprised we were content with NullJSONField returning `None` when used with `values_list` previously. I think generally that `test_save_empty_properties` test confused me since it was hard to determine why we expected the value to return an empty dict in one case but None in another case. Now it is unified which I definitely prefer. Thanks for the change!

**Comment by gherceg** [comment]
Since I can't officially approve my own PR, just noting that I approve of Daniel's commits.

File: corehq/blobs/tests/test_metadata.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/blobs/tests/test_metadata.py]: Yeah, this was my fault. I should not have encoded the (incorrect) `values_list` result in the test. Or better yet, should have implemented `from_db_value` back when I originally wrote that code.

**Review by AmitPhulera — APPROVED** [review]

**Review by jingcheng16 — APPROVED** [review]
